torch/_subclasses/functional_tensor.py

```diff
@@ -0,0 +1,259 @@
+            torch.ops.aten.size.default,
+            torch.ops.aten.sym_size.default,
+            torch.ops.aten.stride.default,
+            torch.ops.aten.sym_stride.default,
+            torch.ops.aten.storage_offset.default,
+            torch.ops.aten.sym_storage_offset.default,
+            torch.ops.aten.numel.default,
+            torch.ops.aten.sym_numel.default,
+            torch.ops.aten.dim.default,
+        ]:
+
+            def unwrap(x):
+                return x.elem
+
+            args_unwrapped = pytree.tree_map_only(FunctionalTensor, unwrap, args)
+            # All metadata accesses should be plumbed to the inner tensor, that way we don't have to worry
+            # about the problem of keeping metadata in sync between the wrapper and inner tensor.
+            # This also alleviates us from having to manually handle metadata mutations on the wrapper.
+            return func(*args_unwrapped, **kwargs)
+        # Originally I tried to implement my subclass without giving it a torch_dispatch, but I gave up:
+        # - _make_wrapper_subclass requires a __torch_dispatch__
+        # - If we want to use _make_subclass(), we have a problem: the subclass will share a TensorImpl with the inner tensor,
+        #   which is of type FunctionalTensorWrapper! We explicitly do not want our wrapper to be a FunctionalTensorWrapper.
+        # - If we use the default tensor.__new__(), we have another problem: it returns inner_tensor.alias(),
+        #   which causes every subclass created above autograd to have autograd view metadata (in addition to also being a FunctionalTensorWrapper).
+        raise RuntimeError(
+            "Attempting to use FunctionalTensor on its own. Instead, please use it with a corresponding FunctionalTensorMode()"
+        )
+
+    def __repr__(self):
+        return f"FunctionalTensor({str(self.elem)})"
+
+    @staticmethod
+    def to_functional(x):
+        # We will do the wrapping for the user.
+        assert not torch._is_functional_tensor(x)
+        # The only autograd metadata we care about on the FunctionalTensor is:
+        # - requires_grad (so autograd runs)
+        # - is_leaf (so that mutations on graph inputs that are not leaves are allowed by the autograd engine)
+        #   this is handled by FunctionalTensor.to_functional
+        x_functional = torch._to_functional_tensor(x, mirror_autograd_meta=True)
+
+        torch._functionalize_enable_reapply_views(True)
+        if x.requires_grad and not x.is_leaf:
+            out = _ToFunctionalTensor.apply(x_functional)
+            return out
+        else:
+            out = FunctionalTensor(x_functional)
+            out.requires_grad = x_functional.requires_grad
+            return out
+
+    def from_functional(self):
+        torch._sync(self)
+        return torch._from_functional_tensor(self.elem)
+
+
+class FunctionalTensorMode(TorchDispatchMode):
+    def __init__(self):
+        self.is_active = True
```

Comment:
Now that we have a slot for these mode keys, my general attitude is that we shouldn't need `is_active` bool flag. You should explain why `is_active` is not equivalent to just removing the mode from its slot entirely

Reply:
You're right - I'll just unset the mode when necessary, thanks!